### PR TITLE
Create AzureDevops Git Extension

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/lib/fields.ts [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+  if (!record && pr?.sourceRefName) {
+    record = await referenceToRecordFromTitle(pr?.sourceRefName?.replace('refs/heads/', '') ?? '');
+  }
+
+  if (record) {
+    await linkPullRequestToRecord(pr, record);
+  }
+
+  return record;
+};
```
Comment (trydionel) [inline src/lib/fields.ts]: @sorza-jimmy This appears to search the branch name only. Could you revise it to search by PR title as well?

File: src/lib/fields.ts [hunk 2]
```diff
@@ -0,0 +1,320 @@
+import { IDENTIFIER } from './extension';
+
+/**
+ * Link Pull Request to Aha Record
+ *
+ * @param PR
+ * @param record
+ */
+export const linkPullRequestToRecord = async (PR: AzureDevops.PR, record: Aha.RecordUnion) => {
```
Comment (trydionel) [inline src/lib/fields.ts:9]: @sorza-jimmy I'm seeing exceptions in my test instance from this method. It seems that something is calling this with a null `record`.

File: src/webhooks/webhook.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import { runCommand } from '@lib/runCommand.js';
+import { linkMergeRequest, linkBranch, referenceToRecordFromTitle, linkPullRequestToRecord } from '../lib/fields.js';
+
+aha.on('webhook', async ({ headers, payload }) => {
+  const event = payload.eventType;
+
+  console.log(`Received webhook '${event}' ${payload.event_type || ''}`);
+
+  switch (event) {
+    case 'git.branch.created':
```
Comment (trydionel) [inline src/webhooks/webhook.ts]: @sorza-jimmy How do you register this service hook in Azure? I can only generate a `git.push` event for new branches.

File: src/webhooks/webhook.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      await handleCreateBranch(payload);
+      break;
+    case 'git.pullrequest.created':
+    case 'git.pullrequest.updated':
```
Comment (trydionel) [inline src/webhooks/webhook.ts:14]: I believe this needs to handle a `git.pullrequest.merged` event as well – I don't see the status changing in Aha! Develop after merging/closing a PR.

**Review by trydionel — CHANGES_REQUESTED** [review]
Great start Jimmy! Found a few issues when testing it though. Could you take a look?

**Comment by ghost** [comment]
@trydionel I will fix soon

File: package.json [hunk 5]
```diff
@@ -0,0 +1,67 @@
+{
+  "name": "aha-develop.azure-devops-git-view",
```
Comment (trydionel) [inline package.json]:
```suggestion
  "name": "aha-develop.azure-repos",
```

File: package.json [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  "description": "Azure Devops Git View",
```
Comment (trydionel) [inline package.json]:
```suggestion
  "description": "Azure Repos integration",
```

File: package.json [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  "version": "0.0.0",
+  "author": "@aha-develop/azure-devops-git-view",
```
Comment (trydionel) [inline package.json]:
```suggestion
  "author": "Aha! (support@aha.io)",
```

File: package.json [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    "prettier": "^2.5.1"
+  },
+  "ahaExtension": {
+    "cspSources": [
+      "https://dev.azure.com"
+    ],
+    "contributes": {
+      "views": {
+        "links": {
+          "title": "AzureDevopsGit",
```
Comment (trydionel) [inline package.json]:
```suggestion
          "title": "Azure Repos",
```

File: package.json [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+          "entryPoint": "src/commands/addLink.ts"
+        },
+        "removeLinks": {
+          "title": "AzureDevops: Remove linked merge requests from record",
+          "entryPoint": "src/commands/removeLinks.ts"
+        }
+      },
+      "endpoints": {
+        "webhook": {
+          "title": "Webhook from Azure Devops Git",
```
Comment (trydionel) [inline package.json]:
```suggestion
          "title": "Webhook from Azure DevOps",
```

File: package.json [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+          "recordTypes": [
+            "Feature",
+            "Epic",
+            "Requirement"
+          ]
+        }
+      },
+      "commands": {
+        "sync": {
+          "title": "AzureDevops: Sync merge requests",
```
Comment (trydionel) [inline package.json]:
```suggestion
          "title": "Azure Repos: Sync pull requests",
```

File: package.json [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+          "entryPoint": "src/commands/sync.ts"
+        },
+        "addLink": {
+          "title": "AzureDevops: Link merge request to record",
```
Comment (trydionel) [inline package.json]:
```suggestion
          "title": "Azure Repos: Link pull request to record",
```

File: package.json [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+          "entryPoint": "src/commands/addLink.ts"
+        },
+        "removeLinks": {
+          "title": "AzureDevops: Remove linked merge requests from record",
```
Comment (trydionel) [inline package.json]:
```suggestion
          "title": "Azure Repos: Remove linked pull requests from record",
```

File: src/components/attribute/Menu.tsx [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+export const Menu = ({ record }: MenuProps) => {
+  return (
+    <aha-menu>
+      <aha-button slot="button" type="attribute" size="small">
+        <aha-icon icon="fa-solid fa-ellipsis"></aha-icon>
+      </aha-button>
+      <aha-menu-item onClick={() => runCommand(record, 'sync')}>Resync</aha-menu-item>
+      <aha-menu-item onClick={() => runCommand(record, 'addLink')}>Link merge request</aha-menu-item>
+      <aha-menu-item onClick={() => runCommand(record, 'removeLinks')}>Unlink merge requests</aha-menu-item>
```
Comment (trydionel) [inline src/components/attribute/Menu.tsx]:
```suggestion
      <aha-menu-item onClick={() => runCommand(record, 'addLink')}>Link pull request</aha-menu-item>
      <aha-menu-item onClick={() => runCommand(record, 'removeLinks')}>Unlink pull requests</aha-menu-item>
```

**Review by trydionel — CHANGES_REQUESTED** [review]
Added some suggestions for copy changes as well

File: src/webhooks/webhook.ts (hunk 3 above)
Comment (ghost) [reply src/webhooks/webhook.ts]: @trydionel  Sorry. I missed that. I updated

File: src/webhooks/webhook.ts (hunk 4 above)
Comment (ghost) [reply src/webhooks/webhook.ts:14]: @trydionel  I could not find the `git.pullrequest.merged` event. I got the merged status via `git.pullrequest.updated` event.
image

**Comment by ghost** [comment]
@trydionel  Fixed.

**Comment by ghost** [comment]
@trydionel  Please let me know if you have other issues. Thank you.

**Review by trydionel — APPROVED** [review]
I looked over this earlier today. All of those fixes look good! I've pinged @justinpaulson to take over from here.

**Comment by ghost** [comment]
Thank you.

File: src/lib/fields.ts [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+};
+
+/**
+ *
+ * @param branchName
+ * @param repoUrl
+ * @param record
+ */
+export const linkBranchToRecord = async (branchName: string, repoUrl: string, record: Aha.RecordUnion) => {
+  console.log('~~~~~~~~' + branchName.split('/'));
```
Comment (justinpaulson) [inline src/lib/fields.ts]: can we remove this log?

File: src/lib/fields.ts [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+ * @param str
+ * @returns
+ */
+export const referenceToRecordFromTitle = async (str: string): Promise<Aha.RecordUnion | null> => {
+  const ahaReference = extractReference(str);
+
+  if (!ahaReference) {
+    return null;
+  }
+  console.log(`Searching for ${ahaReference.type} ref ${ahaReference.referenceNum}`);
```
Comment (justinpaulson) [inline src/lib/fields.ts]: Remove this log as well, and line #269 and #289

File: src/webhooks/webhook.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
+import { runCommand } from '@lib/runCommand.js';
+import { linkPullRequest, linkBranch, referenceToRecordFromTitle, linkPullRequestToRecord } from '../lib/fields.js';
+
+const EMPTY_SHA = '0000000000000000000000000000000000000000';
+
+aha.on('webhook', async ({ headers, payload }) => {
+  const event = payload.eventType;
+
+  console.log(`Received webhook '${event}' ${payload.event_type || ''}`);
```
Comment (justinpaulson) [inline src/webhooks/webhook.ts]: remove log

File: src/webhooks/webhook.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+  switch (event) {
+    case 'git.push':
+      await handleCreateBranch(payload);
+      break;
+    case 'git.pullrequest.created':
+    case 'git.pullrequest.updated':
+      await handlePullRequest(payload);
+      break;
+    default:
+      console.log(`No action registered for ${event}`);
```
Comment (justinpaulson) [inline src/webhooks/webhook.ts]: remove log

File: package.json [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+  "license": "MIT",
+  "scripts": {
+    "format": "prettier --write 'src/**/*.ts'"
+  },
+  "dependencies": {
+    "@aha-app/aha-develop-react": "^1.0.6",
+    "axios": "^0.25.0",
+    "graphql": "^16.2.0",
+    "graphql-request": "^3.7.0",
+    "lodash": "^4.17.21"
```
Comment (justinpaulson) [inline package.json]: is this used?

File: package.json [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+  },
+  "license": "MIT",
+  "scripts": {
+    "format": "prettier --write 'src/**/*.ts'"
+  },
+  "dependencies": {
+    "@aha-app/aha-develop-react": "^1.0.6",
+    "axios": "^0.25.0",
+    "graphql": "^16.2.0",
+    "graphql-request": "^3.7.0",
```
Comment (justinpaulson) [inline package.json]: are these GraphQL dependencies used?

**Review by justinpaulson — CHANGES_REQUESTED** [review]
A couple cleanup items

**Review by justinpaulson — APPROVED** [review]

**Pull request merged**